blockchain/tx_list.go

```diff
@@ -308,30 +308,33 @@ func (l *txList) Overlaps(tx *types.Transaction) bool {
 //
 // If the new transaction is accepted into the list, the lists' cost and gas
 // thresholds are also potentially updated.
-func (l *txList) Add(tx *types.Transaction, priceBump uint64) (bool, *types.Transaction) {
+func (l *txList) Add(tx *types.Transaction, priceBump uint64, kip71Hardforked bool) (bool, *types.Transaction) {
 	// If there's an older better transaction, abort
 	old := l.txs.Get(tx.Nonce())
 	if old != nil {
-		// TODO : Need to KIP-71 hardfork.
-		//if tx.Type().IsCancelTransaction() {
-		//	logger.Trace("New tx is a cancel transaction. replace it!", "old", old.String(), "new", tx.String())
-		//} else {
-		//	logger.Trace("already nonce exist", "nonce", tx.Nonce(), "with gasprice", old.GasPrice(), "priceBump", priceBump, "new tx.gasprice", tx.GasPrice())
-		//	return false, nil
-		//}
-
-		// If gas price of older is bigger than newer, abort.
-		if old.GasPrice().Cmp(tx.GasPrice()) >= 0 {
-			logger.Trace("already nonce exist", "nonce", tx.Nonce(), "with gasprice", old.GasPrice(), "priceBump", priceBump, "new tx.gasprice", tx.GasPrice())
-			return false, nil
+		if kip71Hardforked {
+			// If tx is CancelTransaction, replace it even thought tx has lower gasPrice than previous tx.
+			if tx.Type().IsCancelTransaction() {
```

Comment:
This is not important, but how about making `if tx.Type().IsCancelTransaction()` as a common condition like below?

```
// If tx is CancelTransaction, replace it even thought tx has lower gasPrice than previous tx.
if tx.Type().IsCancelTransaction() {
    logger.Trace("New tx is a cancel transaction. replace it!", "old", old.String(), "new", tx.String())
} else if kip71Hardforked {
    if old.GasPrice().Cmp(tx.GasPrice()) >= 0 {
        // If gas price of older is bigger than newer, abort.
        logger.Trace("already nonce exist", "nonce", tx.Nonce(), "with gasprice", old.GasPrice(), "priceBump", priceBump, "new tx.gasprice", tx.GasPrice())
        return false, nil
    }
    // Otherwise overwrite the old transaction with the current one.
    logger.Trace("The transaction was substituted by competitive gas price", "old", old.String(), "new", tx.String())
} else {
    logger.Trace("already nonce exist", "nonce", tx.Nonce(), "with gasprice", old.GasPrice(), "priceBump", priceBump, "new tx.gasprice", tx.GasPrice())
    return false, nil
}
```

Reply:
It is more clear. refactored it.